cmd/disk-cache.go

```diff
@@ -0,0 +1,764 @@
+	// exclusion pattern that matched
+	if !dcache.diskAvailable(size*cacheSizeMultiplier) || c.isCacheExclude(bucket, object) {
+		return putObjectFn(bucket, object, r, metadata)
+	}
+	objInfo = ObjectInfo{}
+	// Initialize pipe to stream data to backend
+	pipeReader, pipeWriter := io.Pipe()
+	hashReader, err := hash.NewReader(pipeReader, size, r.MD5HexString(), r.SHA256HexString())
+	if err != nil {
+		return ObjectInfo{}, err
+	}
+	// Initialize pipe to stream data to cache
+	rPipe, wPipe := io.Pipe()
+	cHashReader, err := hash.NewReader(rPipe, size, r.MD5HexString(), r.SHA256HexString())
+	if err != nil {
+		return ObjectInfo{}, err
+	}
+	oinfoCh := make(chan ObjectInfo)
+	go func() {
+		oinfo, perr := putObjectFn(bucket, object, hashReader, metadata)
+
+		if perr != nil {
+			pipeWriter.CloseWithError(perr)
+			close(oinfoCh)
+			return
+		}
+		oinfoCh <- oinfo
+
+	}()
+	go func() {
+		if err = dcache.Put(bucket, object, cHashReader, metadata, size); err != nil {
+			wPipe.CloseWithError(err)
+			return
+		}
+	}()
+
+	mwriter := io.MultiWriter(pipeWriter, wPipe)
+	_, err = io.Copy(mwriter, r)
+	if err != nil {
+		pipeWriter.CloseWithError(err)
+		wPipe.CloseWithError(err)
+	}
+	objInfo = <-oinfoCh
+	return objInfo, err
+}
+
+// NewMultipartUpload - Starts a new multipart upload operation to backend and cache.
+func (c cacheObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	newMultipartUploadFn := c.NewMultipartUploadFn
+
+	if c.isCacheExclude(bucket, object) {
+		return newMultipartUploadFn(bucket, object, metadata)
+	}
+
+	dcache, err := c.cache.getCacheFS(bucket, object, false)
+	if err != nil {
+		// disk cache could not be located,execute backend call.
+		return newMultipartUploadFn(bucket, object, metadata)
+	}
+
+	uploadID, err = newMultipartUploadFn(bucket, object, metadata)
+	if err != nil {
+		return
+	}
+	// create new multipart upload in cache with same uploadID
+	dcache.NewMultipartUpload(bucket, object, metadata, uploadID)
+	return uploadID, err
+}
+
+// PutObjectPart - uploads part to backend and cache simultaneously.
+func (c cacheObjects) PutObjectPart(bucket, object, uploadID string, partID int, data *hash.Reader) (info PartInfo, err error) {
+	putObjectPartFn := c.PutObjectPartFn
+	dcache, err := c.cache.getCacheFS(bucket, object, false)
+	if err != nil {
+		// disk cache could not be located,execute backend call.
+		return putObjectPartFn(bucket, object, uploadID, partID, data)
+	}
+
+	size := data.Size()
+	// make sure cache has at least cacheSizeMultiplier * size available
+	if !dcache.diskAvailable(size*cacheSizeMultiplier) || c.isCacheExclude(bucket, object) {
+		return putObjectPartFn(bucket, object, uploadID, partID, data)
+	}
+	//Hold the lock so that disk cache and fs put operations for each part get locks in same order.
+	objectMPartPathLock := dcache.nsMutex.NewNSLock(pathJoin(cacheFSObjectsLockPrefix, minioMetaMultipartBucket), pathJoin(bucket, object))
```

Comment:
you need not hold the lock here as cache will not be shared with other minio instances.